src/io/mouse.js

```diff
@@ -1,20 +1,47 @@
 var MathUtil = require('../util/math-util');
 
-function Mouse () {
+function Mouse (runtime) {
     this._x = 0;
     this._y = 0;
     this._isDown = false;
+    /**
+     * Reference to the owning Runtime.
+     * Can be used, for example, to activate hats.
+     * @type{!Runtime}
+     */
+    this.runtime = runtime;
 }
 
 Mouse.prototype.postData = function(data) {
     if (data.x) {
-        this._x = data.x;
+        this._x = data.x - data.canvasWidth / 2;
     }
     if (data.y) {
-        this._y = data.y;
+        this._y = data.y - data.canvasHeight / 2;
     }
     if (typeof data.isDown !== 'undefined') {
         this._isDown = data.isDown;
+        if (this._isDown) {
+            this._activateClickHats(data.x, data.y);
+        }
+    }
+};
+
+Mouse.prototype._activateClickHats = function (x, y) {
```

Comment:
I think that all makes sense to me. Yeah, I was always thinking of it as a place for "raw data" from a particular device/domain to get processed into "Scratch-consumable-data" and events. Probably I/O device is not the best name here, I'll try to think of a new one sometime...

And agreed re: message passing - it will be good when we resolve a central solution for that or remove the worker.

Video sensing: agreed, passing that across with messages will likely be ugly.
Variables: yeah, the name feels weird there. Almost makes sense if you think of variables as memory/RAM/a hard disk/a Dropbox full of cloud data/etc.
Audio/stage: Yep, not sure. I think the need for that becomes less so if we strip away the worker. With the worker, it seems like we need a place for it.

But yeah all in all I agree. Now regarding this PR...any steps forward? :P